src/test/java/com/google/sps/servlets/StudyServletTest.java

```diff
@@ -1 +1,305 @@
+
+    assertTrue(compareJson(response, expectedResponse));
+  }
+
+  @Test
+  public void userNotLoggedIn() throws Exception {
+    helper.setEnvIsLoggedIn(false);
+    List<Card> noCardsQueried = new ArrayList<>();
+
+    servlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+    String expectedResponse = "";
+
+    assertTrue(compareJson(response, expectedResponse));
+  }
+
+  @Test
+  public void userAnswersCorrectly() throws Exception {
+    // The time gap between when the user created the card and when its being tested is 45 hours
+    cardA =
+        new Card.Builder()
+            .setFamiliarityScore(.5)
+            .setTimeTested(0L)
+            .setImageBlobKey("null")
+            .setRawText("test")
+            .setTextTranslated("translatedTest")
+            .build();
+
+    Entity folder = new Entity("Folder", "testID");
+    String folderKey = KeyFactory.keyToString(folder.getKey());
+
+    Card cardInDatastore = storeCardInDatastore(cardA, datastore, folderKey);
+    String cardKey = cardInDatastore.getCardKey();
+
+    when(mockRequest.getParameter("answeredCorrectly")).thenReturn("true");
+    when(mockRequest.getParameter("cardKey")).thenReturn(cardKey);
+
+    servlet.doPost(mockRequest, mockResponse);
+
+    Entity updatedCard = datastore.get(KeyFactory.stringToKey(cardKey));
+    Card cardAUpdated = new Card(updatedCard);
+    Double response = cardAUpdated.getFamiliarityScore();
+    long time = cardAUpdated.getTimeTested();
+    Double expectedResponse = (Double) 2.142857142;
+
+    assertTrue(compareDoubles(response, expectedResponse));
+  }
+
+  @Test
+  public void userAnswersInCorrectly() throws Exception {
+    // The time gap between when the user created the card and when its being tested is 10 hours
+    cardA =
+        new Card.Builder()
+            .setFamiliarityScore(.5)
+            .setTimeTested(126000000L)
+            .setImageBlobKey("null")
+            .setRawText("test")
+            .setTextTranslated("translatedTest")
+            .build();
+
+    Entity folder = new Entity("Folder", "testID");
+    String folderKey = KeyFactory.keyToString(folder.getKey());
+
+    Card cardInDatastore = storeCardInDatastore(cardA, datastore, folderKey);
+    String cardKey = cardInDatastore.getCardKey();
+
+    when(mockRequest.getParameter("answeredCorrectly")).thenReturn("false");
+    when(mockRequest.getParameter("cardKey")).thenReturn(cardKey);
+
+    servlet.doPost(mockRequest, mockResponse);
+
+    Entity updatedCard = datastore.get(KeyFactory.stringToKey(cardKey));
+    Card cardAUpdated = new Card(updatedCard);
+    Double response = cardAUpdated.getFamiliarityScore();
+    long time = cardAUpdated.getTimeTested();
+    Double expectedResponse = (Double) 0.74;
+    assertTrue(compareDoubles(response, (-1) * expectedResponse));
+  }
+
+  public Card storeCardInDatastore(Card card, DatastoreService datastore, String folderKey) {
```

Comment:
Do we have a static method on Card that does this?

Reply:
No theres only one that creates a new entity or a new card.